src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestSPARQLWithTranslation.java

```diff
@@ -0,0 +1,223 @@
+		final SPARQLUnionPatternImpl unionTranslation = new SPARQLUnionPatternImpl();
+		for (final SPARQLGraphPattern i : up.getSubPatterns()) {
+			if (i instanceof TriplePattern) {
+				unionTranslation.addSubPattern(handleTriplePattern((TriplePattern) i));
+			} else if (i instanceof BGP) {
+				unionTranslation.addSubPattern(handleBGP((BGP) i));
+			} else if (i instanceof SPARQLUnionPattern) {
+				unionTranslation.addSubPattern(handleUnion((SPARQLUnionPattern) i));
+			} else if (i instanceof SPARQLGroupPattern) {
+				unionTranslation.addSubPattern(handleGroup((SPARQLGroupPattern) i));
+			} else {
+				throw new FederationAccessException(i.toString(), req, fm);
+			}
+		}
+		return unionTranslation;
+	}
+	
+	protected SPARQLGraphPattern handleGroup(SPARQLGroupPattern gp) throws FederationAccessException {
+		List<SPARQLGraphPattern> subPatterns = new ArrayList<>();
+		for (final SPARQLGraphPattern i : gp.getSubPatterns()) {
+			if (i instanceof TriplePattern) {
+				subPatterns.add(handleTriplePattern((TriplePattern) i));
+			} else if (i instanceof BGP) {
+				subPatterns.add(handleBGP((BGP) i));
+			} else if (i instanceof SPARQLUnionPattern) {
+				subPatterns.add(handleUnion((SPARQLUnionPattern) i));
+			} else if (i instanceof SPARQLGroupPattern) {
+				subPatterns.add(handleGroup((SPARQLGroupPattern) i));
+			} else {
+				throw new FederationAccessException(i.toString(), req, fm);
+			}
+		}
+		
+		return new SPARQLGroupPatternImpl(subPatterns);
+	}
+	
+	protected SPARQLGraphPattern handleOp(Op op) throws FederationAccessException {
+		if (op instanceof OpJoin) {
+			return handleJoin((OpJoin) op);
+		} else if (op instanceof OpUnion) {
+			return handleOpUnion((OpUnion) op);
+		} else if (op instanceof OpFilter) {
+			return handleOpFilter((OpFilter) op);
+		} else if (op instanceof OpBGP) {
+			return handleOpBGP((OpBGP) op);
+		} else if (op instanceof OpSequence) {
+			return handleOpSequence((OpSequence) op);
+		} else {
+			throw new FederationAccessException("Unsupported type of pattern: " + op.getClass().getName(), req, fm);
+		}
+	}
+	
+	protected SPARQLGraphPattern handleJoin(OpJoin oj) throws FederationAccessException {
+		List<SPARQLGraphPattern> subPatterns = new ArrayList<>();
+		Op left = oj.getLeft();
+		subPatterns.add(handleOp(left));
+		Op right = oj.getRight();
+		subPatterns.add(handleOp(right));
+		return new SPARQLGroupPatternImpl(subPatterns);
+	}
+	
+	protected SPARQLGraphPattern handleOpUnion(OpUnion ou) throws FederationAccessException {
+		final SPARQLUnionPatternImpl unionTranslation = new SPARQLUnionPatternImpl();
+		Op left = ou.getLeft();
+		unionTranslation.addSubPattern(handleOp(left));
+		Op right = ou.getRight();
+		unionTranslation.addSubPattern(handleOp(right));
+		return unionTranslation;
+	}
+	
+	//TODO: Handle OpFilter
+	protected SPARQLGraphPattern handleOpFilter(OpFilter of) throws FederationAccessException {
+		return null;
```

Comment:
No!! Simply returning `null` here is a guarantee for giving us headaches later when we all of a sudden have an unexpected `null` somewhere in a rewritten query pattern. To make sure we will always know that there was an unhandled case, throw an exception here (or remove this method and the corresponding case in the `handleOp(Op)` function).